prov/efa/src/efa_tostr.c

```diff
@@ -0,0 +1,109 @@
+	static const char *errno_str[] = {
+		[FI_EFA_ERR_OTHER                   - FI_EFA_ERRNO_OFFSET] = UNKNOWN_ERR_STR,
+		[FI_EFA_ERR_DEPRECATED_PKT_TYPE     - FI_EFA_ERRNO_OFFSET] = "Deprecated packet type encountered",
+		[FI_EFA_ERR_INVALID_PKT_TYPE        - FI_EFA_ERRNO_OFFSET] = "Invalid packet type encountered",
+		[FI_EFA_ERR_UNKNOWN_PKT_TYPE        - FI_EFA_ERRNO_OFFSET] = "Unknown packet type encountered",
+		[FI_EFA_ERR_PKT_POST                - FI_EFA_ERRNO_OFFSET] = "Failure to post packet",
+		[FI_EFA_ERR_PKT_POST_OR_QUEUE       - FI_EFA_ERRNO_OFFSET] = "Failure to post or queue packet",
+		[FI_EFA_ERR_PKT_SEND_QUEUED         - FI_EFA_ERRNO_OFFSET] = "Encountered error sending queued packet",
+		[FI_EFA_ERR_PKT_PROC_MSGRTM         - FI_EFA_ERRNO_OFFSET] = "Error processing non-tagged RTM",
+		[FI_EFA_ERR_PKT_PROC_TAGRTM         - FI_EFA_ERRNO_OFFSET] = "Error processing tagged RTM",
+		[FI_EFA_ERR_PKT_ALREADY_PROCESSED   - FI_EFA_ERRNO_OFFSET] = "Packet already processed",
+		[FI_EFA_ERR_OOM                     - FI_EFA_ERRNO_OFFSET] = "Out of memory",
+		[FI_EFA_ERR_MR_CLOSE                - FI_EFA_ERRNO_OFFSET] = "MR close error",
+		[FI_EFA_ERR_MR_DEREG                - FI_EFA_ERRNO_OFFSET] = "MR deregistration error",
+		[FI_EFA_ERR_RX_ENTRY_COPY           - FI_EFA_ERRNO_OFFSET] = "RX entry copy error",
+		[FI_EFA_ERR_RX_ENTRIES_EXHAUSTED    - FI_EFA_ERRNO_OFFSET] = "RX entries exhausted",
+		[FI_EFA_ERR_TX_ENTRIES_EXHAUSTED    - FI_EFA_ERRNO_OFFSET] = "TX entries exhausted",
+		[FI_EFA_ERR_AV_INSERT               - FI_EFA_ERRNO_OFFSET] = "Failure inserting address into address vector",
+		[FI_EFA_ERR_RMA_ADDR                - FI_EFA_ERRNO_OFFSET] = "RMA address verification failed",
+		[FI_EFA_ERR_BULK_POST_RECV_BUF      - FI_EFA_ERRNO_OFFSET] = "Bulk post internal receive buffers",
+		[FI_EFA_ERR_PEER_HANDSHAKE          - FI_EFA_ERRNO_OFFSET] = "Failure to post handshake to peer",
+		[FI_EFA_ERR_POST_FLUSH              - FI_EFA_ERRNO_OFFSET] = "Failure to post work request to queue pair",
+		[FI_EFA_ERR_RTM_MISMATCH            - FI_EFA_ERRNO_OFFSET] = "RTM size mismatch",
+		[FI_EFA_ERR_READ_POST               - FI_EFA_ERRNO_OFFSET] = "Error posting read request",
+		[FI_EFA_ERR_RDMA_READ_POST          - FI_EFA_ERRNO_OFFSET] = "Error posting RDMA read request",
+		[FI_EFA_ERR_RDMA_READ_POST_OR_QUEUE - FI_EFA_ERRNO_OFFSET] = "Error posting or queueing RDMA read request",
+		[FI_EFA_ERR_INVALID_DATATYPE        - FI_EFA_ERRNO_OFFSET] = "Invalid datatype encountered",
+		[FI_EFA_ERR_WRITE_COMP              - FI_EFA_ERRNO_OFFSET] = "Failure to write completion",
+		[FI_EFA_ERR_SHM_INTERNAL_ERROR      - FI_EFA_ERRNO_OFFSET] = "SHM internal error",
+		[FI_EFA_ERR_SHM_CQ_ENTRY            - FI_EFA_ERRNO_OFFSET] = "Failure to write CQ entry for SHM operation",
+	};
+
+	return (err < FI_EFA_ERRNO_OFFSET || err >= FI_EFA_ERRNO_MAX)
+		? UNKNOWN_ERR_STR
+		: errno_str[err - FI_EFA_ERRNO_OFFSET];
+}
+
+const char *efa_strerror(int prov_errno)
```

Comment:
Suggest documenting this function, i.e. `prov_errno` must be positive.

Reply:
+1 add doxygen style formatting to your functions that explain input, output, valid arguments, summary, etc.

Reply:
You can also add an assert(prov_errno >= 0)